rules/onchain_budget.go

```diff
@@ -0,0 +1,687 @@
+		reqBytes, err := store.Get(ctx, reqPref)
+		if err != nil {
+			return err
+		}
+
+		// Some requests can have multiple responses. So if we can't
+		// find an entry in the in-mem store corresponding to the given
+		// request ID, then we assume it has already been accounted for.
+		if len(reqBytes) == 0 {
+			return nil
+		}
+
+		// If we did find an entry, we can now delete it.
+		err = store.Del(ctx, reqPref)
+		if err != nil {
+			return err
+		}
+
+		var request onChainState
+		if err := json.Unmarshal(reqBytes, &request); err != nil {
+			return err
+		}
+
+		// Now, we subtract the reqAmt from total pending amt and store
+		// it.
+		if pending.Amount < request.Amount {
+			return fmt.Errorf("total pending cannot be less than " +
+				"requested amount")
+		}
+		pending.Amount -= request.Amount
+
+		b, err := json.Marshal(pending)
+		if err != nil {
+			return err
+		}
+
+		err = store.Set(ctx, onChainPending, b)
+		if err != nil {
+			return err
+		}
+
+		// Finally, we add the amount to the total spent and persist it
+		// to the permanent store.
+		complete.Amount += request.Amount
+		b, err = json.Marshal(complete)
+		if err != nil {
+			return err
+		}
+
+		return store.Set(ctx, onChainSpent, b)
+	})
+}
+
+// getBudgetState fetches the current state of the budget by getting the total
+// amount persisted store along with the total pending amount from the temporary
+// store.
+func (o *OnChainBudgetEnforcer) getBudgetState(ctx context.Context,
+	tx firewalldb.KVStoreTx) (*onChainState, *onChainState, error) {
+
+	var (
+		spentAmt   onChainState
+		pendingAmt onChainState
+	)
+
+	store := tx.Local()
+
+	spentBytes, err := store.Get(ctx, onChainSpent)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pendingBytes, err := store.Get(ctx, onChainPending)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(spentBytes) != 0 {
+		err := json.Unmarshal(spentBytes, &spentAmt)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if len(pendingBytes) != 0 {
+		err := json.Unmarshal(pendingBytes, &pendingAmt)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return &spentAmt, &pendingAmt, nil
+}
+
+// addPendingAmt updates the temporary KV store with the new amount. The total
+// pending amount is updated, and a request specific entry is added so that
+// we know how much to adjust our budget by if this request where to succeed or
+// fail.
+func (o *OnChainBudgetEnforcer) addPendingAmt(ctx context.Context,
+	tx firewalldb.KVStoreTx, pending *onChainState,
+	request *onChainAction) error {
+
+	store := tx.Local()
+
+	b, err := json.Marshal(
+		&onChainState{Amount: pending.Amount + request.Amount},
+	)
+	if err != nil {
+		return err
+	}
+
+	err = store.Set(ctx, onChainPending, b)
+	if err != nil {
+		return err
+	}
+
+	reqID := o.GetReqID()
+	reqPref := fmt.Sprintf("%s:%d", onChainPending, reqID)
+
+	b, err = json.Marshal(request)
+	if err != nil {
+		return err
+	}
+
+	return store.Set(ctx, reqPref, b)
```

Comment:
nit:
Even though lnd should never create another request with the same request ID, should we check if an entry of this key already exists in the Store? Given that there will be pretty bad consequences if we'd overwrite an already existing entry.

Reply:
Ah, ok, understand now better how the request id is created (by lnd), added a check.